src/Foundatio.Storage.SshNet/Storage/SshNetFileStorage.cs

```diff
@@ -13,384 +14,432 @@
-            };
-        }
+        var list = new List<FileSpec>();
+        var criteria = GetRequestCriteria(searchPattern);
 
-        private async Task<List<FileSpec>> GetFileListAsync(string searchPattern = null, int? limit = null, int? skip = null, CancellationToken cancellationToken = default) {
-            if (limit is <= 0)
-                return new List<FileSpec>();
+        EnsureClientConnected();
 
-            var list = new List<FileSpec>();
-            var criteria = GetRequestCriteria(searchPattern);
+        // NOTE: This could be very expensive the larger the directory structure you have as we aren't efficiently doing paging.
+        int? recordsToReturn = limit.HasValue ? skip.GetValueOrDefault() * limit + limit : null;
+        await GetFileListRecursivelyAsync(criteria.Prefix, criteria.Pattern, list, recordsToReturn, cancellationToken).AnyContext();
 
-            EnsureClientConnected();
+        if (skip.HasValue)
+            list = list.Skip(skip.Value).ToList();
 
-            // NOTE: This could be very expensive the larger the directory structure you have as we aren't efficiently doing paging.
-            int? recordsToReturn = limit.HasValue ? skip.GetValueOrDefault() * limit + limit : null;
-            await GetFileListRecursivelyAsync(criteria.Prefix, criteria.Pattern, list, recordsToReturn, cancellationToken).AnyContext();
+        if (limit.HasValue)
+            list = list.Take(limit.Value).ToList();
 
-            if (skip.HasValue)
-                list = list.Skip(skip.Value).ToList();
+        return list;
+    }
 
-            if (limit.HasValue)
-                list = list.Take(limit.Value).ToList();
+    private async Task GetFileListRecursivelyAsync(string prefix, Regex pattern, ICollection<FileSpec> list, int? recordsToReturn = null, CancellationToken cancellationToken = default) {
+        using var _ = _logger.BeginScope(s => s.Property("Prefix", prefix).Property("Pattern", pattern.ToString()));
 
-            return list;
+        _logger.LogTrace("Getting file list recursively matching {Prefix}...", prefix);
+        if (cancellationToken.IsCancellationRequested) {
+            _logger.LogDebug("Returning empty file list: Cancellation requested");
+            return;
         }
 
-        private async Task GetFileListRecursivelyAsync(string prefix, Regex pattern, ICollection<FileSpec> list, int? recordsToReturn = null, CancellationToken cancellationToken = default) {
-            _logger.LogInformation($"Checking {prefix}...");
-            if (cancellationToken.IsCancellationRequested)
-                return;
+        var files = new List<SftpFile>();
+        try {
+            files.AddRange(await _client.ListDirectoryAsync(prefix, null).AnyContext());
+        } catch (SftpPathNotFoundException) {
+            _logger.LogDebug("Returning empty file list: Directory not found with {Prefix}", prefix);
```

Comment:
Again, this is a recursive call and we might have already returned some files.

Reply:
Yes, but we can follow the call stack and see what it did in every folder.